moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp

```diff
@@ -268,14 +268,37 @@ bool MoveItCpp::execute(const std::string& group_name, const robot_trajectory::R
   // Execute trajectory
   moveit_msgs::RobotTrajectory robot_trajectory_msg;
   robot_trajectory->getRobotTrajectoryMsg(robot_trajectory_msg);
-  // TODO: cambel
-  // blocking is the only valid option right now. Add non-bloking use case
+  moveit_controller_manager::ExecutionStatus execution_status;
+
+  auto callback = trajectory_execution_manager::TrajectoryExecutionManager::ExecutionCompleteCallback();
+
+  if (trajectory_execution_manager_->getEnableSimultaneousExecution())
+  {
+    callback = [this,
+                &execution_status](__attribute__((unused)) const moveit_controller_manager::ExecutionStatus status) {
+      execution_complete_condition_.notify_all();
+      execution_status = status;
+    };
+  }
+
+  if (!trajectory_execution_manager_->push(robot_trajectory_msg, "", callback))
+    return false;
+
   if (blocking)
   {
-    trajectory_execution_manager_->push(robot_trajectory_msg);
-    trajectory_execution_manager_->execute();
-    return trajectory_execution_manager_->waitForExecution();
+    if (trajectory_execution_manager_->getEnableSimultaneousExecution())
+    {  // wait for callback to return
+      std::unique_lock<std::mutex> ulock(execution_complete_mutex_);
+      execution_complete_condition_.wait(ulock);
```

Comment:
```suggestion
      execution_complete_condition_.wait(ulock, active_trajectories_.find(trajectory_id) == active_trajectories_.end());
```
There are some problems here:
 - spurious wakeups can happen before notify_all is called  (yes, sounds like a bug ... ) so we need to handle those
 - there could be several blocking trajectories from different threads

Reply:
okay

Reply:
Okay